"use strict";(self["webpackChunknew_mirna"]=self["webpackChunknew_mirna"]||[]).push([[525],{3744:function(e,s){s.Z=(e,s)=>{const t=e.__vccOpts||e;for(const[i,n]of s)t[i]=n;return t}},8933:function(e,s,t){t.r(s),t.d(s,{default:function(){return f}});var i=t(6252),n=t.p+"img/CGMMRC miRNA-seq pipeline.6d7c337d.png";const r=(0,i._)("h1",{class:"mb-5 mx-3 workflowtitleStyle"},"Micro RNA-seq analysis method ",-1),a=(0,i._)("p",{class:"mb-5 ml-3 subscribeText"}," We processed micro RNA-seq data using the Partek Flow analysis software. Initially, we removed low-quality bases and sequencing adapters from the raw FASTQ files. The trimmed reads were then aligned to the whole genome reference using Bowtie 2. Next, we quantified reads assigned to the miRNA listed in the miRBase (v22) and normalized them to CPM (Counts Per Million). Statistical analysis, hierarchical clustering, and differential expression analysis were performed using the GSA (Gene Specific Analysis) or DESeq2 function in Partek Flow. We further analyzed the results using Qiagen IPA (Ingenuity Pathway Analysis) to acquire target genes of each sequenced miRNA. The target genes derived from the differentially expressed miRNAs can be further explored for their potential role in biological processes or diseases. ",-1),l=(0,i._)("div",{class:"d-flex justify-center mt-5"},[(0,i._)("img",{src:n,style:{width:"40vw"}})],-1);function o(e,s){const t=(0,i.up)("v-col"),n=(0,i.up)("v-row"),o=(0,i.up)("v-container");return(0,i.wg)(),(0,i.iD)("div",null,[(0,i.Wm)(o,null,{default:(0,i.w5)((()=>[(0,i.Wm)(n,{class:"d-flex justify-center"},{default:(0,i.w5)((()=>[(0,i.Wm)(t,{cols:"8"},{default:(0,i.w5)((()=>[r,a])),_:1}),(0,i.Wm)(t,{cols:"8"},{default:(0,i.w5)((()=>[l])),_:1})])),_:1})])),_:1})])}var c=t(3744);const d={},u=(0,c.Z)(d,[["render",o]]);var f=u}}]);
//# sourceMappingURL=525.4544ef9a.js.map